go/store/prolly/tree/json_indexed_document.go

```diff
@@ -0,0 +1,233 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tree
+
+import (
+	"context"
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/types"
+	"sync"
+)
+
+type jsonLocationKey = []byte
+
+type address = []byte
+
+type StaticJsonMap = StaticMap[jsonLocationKey, address, jsonLocationOrdering]
+
+// IndexedJsonDocument is an implementation of sql.JSONWrapper that stores the document in a prolly tree.
+// Every leaf node in the tree is a blob containing a substring of the original document. This allows the document
+// to be reconstructed by walking the tree and concatenating the values in the leaves.
+// Every non-leaf node is an address map, using a jsonLocation as the key. This key corresponds to the final nameable
+// location within the JSON substring.
+// The |interfaceFunc| field caches the result of ToInterface in the event it gets called multiple times.
+type IndexedJsonDocument struct {
+	m             StaticJsonMap
+	interfaceFunc func() (interface{}, error)
+}
+
+var _ types.JSONBytes = IndexedJsonDocument{}
+var _ types.MutableJSON = IndexedJsonDocument{}
+var _ types.SearchableJSON = IndexedJsonDocument{}
+var _ fmt.Stringer = IndexedJsonDocument{}
+var _ driver.Valuer = IndexedJsonDocument{}
+
+func NewIndexedJsonDocument(root Node, ns NodeStore) IndexedJsonDocument {
+	m := StaticMap[jsonLocationKey, address, jsonLocationOrdering]{
+		Root:      root,
+		NodeStore: ns,
+		Order:     jsonLocationOrdering{},
+	}
+	return IndexedJsonDocument{
+		m: m,
+		interfaceFunc: sync.OnceValues(func() (interface{}, error) {
+			return getInterfaceFromIndexedJsonMap(m)
+		}),
+	}
+}
+
+// Clone implements types.MutableJSON. Mutating an IndexedJsonDocument always returns a new IndexedJsonDocument without
+// modifying the state of the original. But creating a new instance allows callers to modify the value returned by ToInterface()
+// on the "mutable" copy without affecting the value cached on the original.
+func (i IndexedJsonDocument) Clone() types.MutableJSON {
+	// TODO: Add context parameter to MutableJSON.Clone()
```

Comment:
Yeah. I think I was punting on this until I have everything else working because I'd have to change a GMS API. I'll resolve this comment once I make that change.

Reply:
I added a context parameter to clone, since it's significantly more tractable than adding a context to ToInterface. It's not ideal since while it works now, in the future ToInterface could be called with a different context. But it's good enough for now.